Add QSTK Sale

antho1404 opened the pull request:
fix https://github.com/Quiver-Protocol/emotional-nft-dapp/issues/277

This PR adds both sales (QSTK seed and NFT) to the investor space.
It also connects to the sales smart contract to activate or not the sale button.
When the user wants to purchase QSTK they need to:
- Fill in the amount of token they want to purchase (and visualize the price it will cost them with a simulator)
- Connect their crypto wallet (metamask or wallet connect)
- Select the correct network
- Sign the transaction
- Wait for the transaction to be executed

This flow is nice but requires 2 different connectors that make things quite complicated as one public is needed to calculate the price and the other (called 'user') required to actually purchase based on the user's wallet. Trying to make this flow with a single provider has a lot of side effects like not opening metamask as the public connector is already activated.

antho1404 commented on contract.ts:
```diff
@@ -0,0 +1,160 @@
+export const address = '0xC3252F3f61329CfbcE80fF80D3a156B7c9170884'
```
TODO: use mainnet address

antho1404 commented on providers.ts:
```diff
@@ -0,0 +1,15 @@
+import { WalletConnectConnector } from '@web3-react/walletconnect-connector'
+
+const chains = {
+  3: 'https://eth-ropsten.alchemyapi.io/v2/j3511RMZjDGkirYD0QPu8nGn1sIY0Y7c',
```
TODO: add mainnet setup

antho1404 commented:
Issue on netlify 

```
4:35:56 PM: ModuleNotFoundError: Module not found: Error: Can't resolve 'bufferutil' in '/opt/build/repo/node_modules/ws/lib'
```

I remember this issue a few times on different project but still don't remember how to fix it correctly @NicolasMahe do you remember

NicolasMahe commented on contract.ts:
```diff
@@ -0,0 +1,160 @@
+    stateMutability: 'payable',
+    type: 'receive',
+  },
+]
```
You could only keep the ABI used to make it a bit smaller

NicolasMahe commented:
- [ ] Can you block the min value with `min="1"` of the input?

<img width="431" alt="Screenshot 2021-06-14 at 10 12 48" src="https://user-images.githubusercontent.com/5823445/121834466-153a5700-ccf9-11eb-8c0f-1fcadbf999ab.png">


- [ ] Metamask button doesn't  do anything when clicking it
- [ ] Add loader to WalletConnect button

antho1404 commented:
PR ready for review. It was more complicated than anticipated with the current flow where the user connects the wallet at the last moment

antho1404 commented on pages/_app.tsx:
```diff
@@ -1,18 +1,31 @@
 import ScrollButton from '../components/scroll-button/scroll-button'
 import '../styles/globals.css'
 
+const Web3ReactProviderUser = dynamic(() => import('../components/web3-user'), {
```
This is required to load dynamically as web3react has issues with the SSR of next
https://github.com/NoahZinsmeister/web3-react/issues/176

antho1404 commented on components/web3-user.tsx:
```diff
@@ -0,0 +1,13 @@
+import { createWeb3ReactRoot } from '@web3-react/core'
```
Component loaded dynamically to fix an issue with web3react
https://github.com/NoahZinsmeister/web3-react/issues/176

NicolasMahe commented:
The flow is working well. I was able to buy, but also switch network at random moment and didn't have an issue to continue the flow.

Few feedback on the UI that could be improved:
- I really don't like the close button to be at the top right on the website. It's so far away it took me too much time to find it.
  - I think the button to close is a better UX that closing the modal when clicking on the background and lose the current state 👍 
  - Maybe just move back the button closer to the modal or even inside in the top right corner. I think @EmmanuelDrouin will have a strong advise on this one

- There is no error when opening the modal and click on the purchase button. I was expecting to see a error like "please fill this input" after clicking on the button

- There is no hover border or background effect on any of the buttons. This is should be fixed.

- Metamask and wallet connect icons are missing ;)

NicolasMahe commented on components/templates/inverstor/investor.tsx:
```diff
@@ -1,155 +1,228 @@
+  const [tx, setTx] = useState<ContractTransaction>()
+  const [receipt, setReceipt] = useState<ContractReceipt>()
+
+  const { activate, chainId: sellChainId } = useWeb3React<Web3Provider>()
```
`sellChainId` is not a great name and it should be spelled `saleChainId`.
Maybe focusing on the user vs remote provider? eg:
```suggestion
  const { activate, chainId: remoteChainId } = useWeb3React<Web3Provider>()
```
and the one from the user should also be renamed to `userChainId` to be super explicit

EmmanuelDrouin reviewed (commented):
The flow is good!

Few feedback:

- We cannot click outside the modal to close it, we need to click on the cross to do so. Which is on the top right corner (far way ^^), it was a bit confusing at the end after I bought my tokens and try to close this modal.
- Buying a negative amount:
  - The error is only displayed when I try to sign the transaction to purchase the tokens. It would be better to block it in the input field directly so the user cannot input a negative number 
  - <img width="543" alt="Screen Shot 2021-06-14 at 17 58 48" src="https://user-images.githubusercontent.com/15095289/121882456-c1506200-cd3a-11eb-8be3-0fe9daffc0d3.png">
- The flow is missing "transition" between modals, it really obvious when the user clicks on the "buy token" button and the modal opens for the first time. The effect is not really nice without transition property.

https://user-images.githubusercontent.com/15095289/121883386-e1ccec00-cd3b-11eb-8189-60986b5393cf.mov

  - Example of a nice transition effect for a modal:

https://user-images.githubusercontent.com/15095289/121883564-1476e480-cd3c-11eb-9cc3-a896cad848cd.mov

NicolasMahe commented on components/templates/inverstor/steps.tsx:
```diff
@@ -0,0 +1,210 @@
+  <>
+    <Body2 className="mt-6">Transaction Succeeded</Body2>
+    <Dialog.Description className="mt-2 text-sm leading-5 font-normal text-gray-500 truncate">
+      You successfully purchased {amount.toString()} QSTK tokens
```
could be nice to display the amount a bit better with https://docs.ethers.io/v5/api/utils/display-logic/#utils-commify

NicolasMahe commented on components/templates/inverstor/investor.tsx:
```diff
@@ -1,155 +1,228 @@
+  }, [library, account])
+
+  useEffect(() => {
+    void activate(networkConnector, null, true)
```
if I remember correctly the third argument `true` will make this function to throw error instead of putting the error in the `error` state variable from the same hook as the `activate` function.
It may be much simpler to manage error. don't you think?

gopherquiv commented:
@antho1404 @NicolasMahe I can't close the "Transaction Succeeded" modal after purchase.

antho1404 commented:
I updated a few things. The button is now disabled, so you cannot click if you don't have the right amount. I also fixed the animation for the popup.
The hover on the button will be done on another issue https://github.com/Quiver-Protocol/quiver-website/issues/23.
And I will update the modal to be closable when clicking on the background and/or the close button that will be closer.
